Detectors/MUON/MID/Raw/src/RawBuffer.cxx

```diff
@@ -144,29 +162,16 @@ bool RawBuffer<T>::hasNext(unsigned int nBytes)
   // With lot of memory left in the buffer but no payload
   nextPayload();
   bool isOk = mCurrentBuffer.size() + mUnconsumed.size() - mNextHeaderIndex + mEndOfPayloadIndex - mElementIndex >= nBytes / mElementSizeInBytes;
-  if (!isOk) {
+  if (!isOk && mElementIndex != mCurrentBuffer.size()) {
     // Store the remaining bits for further use
-    // We need to do it here because we know that the vector of which the mBytes is just a spas, is valid
-    // If we do it in set buffer, this might not be the case anymore
-    std::copy(mBytes.begin() + mElementIndex, mBytes.end(), std::back_inserter(mUnconsumed));
+    // We need to do it here because the vector of which the mBytes is just a span might not be valid afterwards
+    // (e.g. when we do the next setBuffer)
+    // If we do not want to invalidate the mRDH pointer, we need to copy bytes from the last header
+    std::copy(mBytes.begin() + mHeaderIndex, mBytes.end(), std::back_inserter(mUnconsumed));
```

Comment:
I've actually not timed the difference myself, but you might also consider writing it like :  

mUnconsumed.insert(mUnconsumed.end(),mBytes.begin()+mHeaderIndex,mBytes.end()) 

which, irrespective of the performance gain (if at all), is slightly more readable IMO.

Reply:
Looking [here](https://stackoverflow.com/questions/14382228/is-inserting-in-the-end-equivalent-to-stdcopy), it is not perfectly clear what one could gain. But the modification is small, so I can perform it.